### ISSUE TITLE
shared_promise is not available immediately after set_value/exception if it has outstanding shared futures

### ISSUE THREAD

**Issue by bhalevy** [scylladb/seastar#1926]
This issue is derived from https://github.com/scylladb/scylladb/issues/15956

The following example fails:
```c++
SEASTAR_THREAD_TEST_CASE(test_shared_promise_with_outstanding_future_is_immediately_available) {
    shared_promise<> pr;
    auto f = pr.get_shared_future();
    pr.set_value();
    BOOST_REQUIRE(pr.available());
    BOOST_REQUIRE_NO_THROW(f.get());
}
```

while, as @piodul noticed, adding a `yield()` after `set_value` allows the state to be set properly:
```c++
SEASTAR_THREAD_TEST_CASE(test_shared_promise_with_outstanding_future_is_immediately_available) {
    shared_promise<> pr;
    auto f = pr.get_shared_future();
    pr.set_value();
    yield().get();
    BOOST_REQUIRE(pr.available());
    BOOST_REQUIRE_NO_THROW(f.get());
}
```

See @piodul's comment https://github.com/scylladb/scylladb/issues/15956#issuecomment-1795332929:
> The only missing detail is that `shared_state::resolve` is called *in a continuation* which is attached to `_original_future`. The `_original_future` waits on a seastar promise inside `shared_promise`, and `shared_promise::set_value` sets that promise. However, setting a seastar promise does not cause the continuation attached to the future to run immediately, it is only `seastar::schedule`d to run later, at the end of the queue of the current scheduling group. Therefore, `available()` does not become true immediately after `set_value`.
>
> If, after the task which calls `set_value`, another task which calls `available()` is already scheduled, then the other task will also see the shared future as unavailable.

**Comment by bhalevy** [comment]
Reopened by 584d2cd80af3f5ddbc8b04b847ff183b3c7d33d1
